src/main/java/hudson/matrix/MatrixProject.java

```diff
@@ -199,14 +201,35 @@
      */
     private String childCustomWorkspace;
 
+    /**
+     * Lock to prevent project changes on different build at the same time
+     */
+    private ReadWriteLock buildLock;
```

Comment:
iff we need this, it should be transient so it is not persisted and initialized in `readResolve` so old configuration have it configured. As implemented it would cause NPEs when old jobs are build.